File: src/services/share.ts
import { invoke } from "@tauri-apps/api/core";

export interface SharedImagePayload {
  file_name: string;
  file_path: string;
  mime_type: string;
  size_bytes: number;
  timestamp: number;
  source: string;
  bytes?: number[];
}

/**
 * Checks if there is any pending shared image waiting from the iOS Share Extension
 * or system share mechanism.
 */
export async function getPendingSharedImage(includeBytes = true): Promise<SharedImagePayload | null> {
  try {
    return await invoke<SharedImagePayload | null>("get_pending_shared_image", {
      includeBytes,
    });
  } catch (err) {
    console.warn("Failed to check pending shared image from native bridge:", err);
    return null;
  }
}

/**
 * Clears pending shared images in the shared container once processed.
 */
export async function clearPendingSharedImage(): Promise<void> {
  try {
    await invoke("clear_pending_shared_image");
  } catch (err) {
    console.warn("Failed to clear pending shared image:", err);
  }
}

/**
 * Stages an image for simulation, dev testing, or local cross-platform share flow.
 */
export async function stageSharedImage(
  bytes: Uint8Array | number[],
  fileName: string,
  mimeType?: string
): Promise<SharedImagePayload> {
  const payloadBytes = bytes instanceof Uint8Array ? Array.from(bytes) : bytes;
  return await invoke<SharedImagePayload>("stage_shared_image", {
    bytes: payloadBytes,
    fileName,
    mimeType: mimeType || "image/png",
  });
}

/**
 * Converts a SharedImagePayload with bytes into a standard browser File object.
 */
export function payloadToFile(payload: SharedImagePayload): File | null {
  if (!payload.bytes || payload.bytes.length === 0) {
    return null;
  }

  const u8Array = new Uint8Array(payload.bytes);
  const blob = new Blob([u8Array], { type: payload.mime_type || "image/png" });
  return new File([blob], payload.file_name, {
    type: payload.mime_type || "image/png",
    lastModified: payload.timestamp ? payload.timestamp * 1000 : Date.now(),
  });
}

File: src-tauri/src/commands/share.rs
use crate::error::AppError;
use crate::share::{self, SharedImagePayload};

#[tauri::command]
pub fn get_pending_shared_image(
    include_bytes: Option<bool>,
) -> Result<Option<SharedImagePayload>, AppError> {
    share::get_pending_shared_image(include_bytes.unwrap_or(true))
}

#[tauri::command]
pub fn clear_pending_shared_image() -> Result<(), AppError> {
    share::clear_pending_shared_image()
}

#[tauri::command]
pub fn stage_shared_image(
    bytes: Vec<u8>,
    file_name: String,
    mime_type: Option<String>,
) -> Result<SharedImagePayload, AppError> {
    share::stage_shared_image(&bytes, &file_name, mime_type.as_deref())
}

#[tauri::command]
pub fn load_image_from_path(path: String) -> Result<SharedImagePayload, AppError> {
    share::load_image_from_path(&path)
}
